andi/andi.py

```diff
@@ -53,10 +59,161 @@ def to_provide(
+    for each type so that the dependencies existence is assured. The keys in
+    this plan are the type, and the values are dicts where the keys are the
+    params names and the values are the type required for this parameter.
+
+    If the input is a function then there will be a final entry in the plan with
+    the key ``FunctionArguments`` that contains the dictionary of the
+    fulfilled arguments for this function (could be incomplete).
+
+    This function recursively checks for dependencies. If a cyclic dependency is
+    found the error ``CyclicDependencyError`` is raised.
+
+    :param class_or_func: If a class is provided this function will create
+                          a plan to fulfil its ``__init__`` method. The class
+                          itself will be part of the plan in the las position.
+                          If a plan for all the requirements cannot be created
+                          the function will fail with ``NonProvidableError``.
+                          If a method is given then this function will try
+                          to create a plan to create all its arguments, but it
+                          won't fail if all the required parameters cannot
+                          be fulfilled. The last entry in this case will be
+                          the arguments that could be fulfilled for the input
+                          function with its types and the key will be
+                          ``FunctionArguments``
+    :param can_provide: A predicate or a dictionary that says if a class
+                        is providable. Any required class found
+                        by this function should be providable, otherwise,
+                        ``NonProvidableError`` will be raised. There one single
+                        exception for that: if a function is received as
+                        input then non providable arguments
+                        are allowed for the function itself. They won't be
+                        in the returned plan. It is the caller responsibility
+                        to deal with this case.
+    :param externally_provided: A predicate or a dictionary that says if a class
+                                will be provided externally.
+                                The ``plan`` function won't try to resolve its
+                                dependencies, so it acts as a way to stop
+                                dependency injection for these classes where
+                                we don't want it because they will be provided by
+                                other means.
+    :return: The plan ready to be used as ``build`` method input.
+    """
+    assert can_provide is not None
+    assert externally_provided is not None
+    if isinstance(can_provide, Container):
+        can_provide = can_provide.__contains__
+    if isinstance(externally_provided, Container):
+        externally_provided = externally_provided.__contains__
```

Comment:
What do you think about adding a helper function (`ensure_can_provide_func`?), which receives TypeContainerOrCallable, and always returns a callable?

Reply:
Done in https://github.com/scrapinghub/andi/pull/8/commits/82b6fb7332e925d76c070bdd84ddf5ce78318832